core/claims/claim.go

```diff
@@ -131,35 +111,187 @@ func NewClaimFromEntry(e *merkletree.Entry) (merkletree.Entrier, error) {
+}
+
+// Unmarshal the ClaimHeader from an entry
+func (c *ClaimHeader) Unmarshal(e *merkletree.Entry) {
+	index := e.Index()
+	copy(c.Type[:], index[0][:ClaimTypeLen])
+	flags0 := index[0][ClaimTypeLen]
+	c.Dest = ClaimRecip(flags0 & 0b00000011)
+	c.Expiration = byte2bool(flags0 & (1 << 2))
+	c.Version = byte2bool(flags0 & (1 << 3))
+}
+
+var (
+	// ClaimHeaderBasic is a simple claim type that can be used for anything.
+	ClaimHeaderBasic = ClaimHeader{
+		Type:       *ClaimTypeBasic,
+		Dest:       ClaimRecipSelf,
+		Expiration: false,
+		Version:    false}
+	// ClaimTypeAuthorizeKSignBabyJub is a claim type to autorize a babyjub public key for signing.
+	ClaimHeaderAuthorizeKSignBabyJub = ClaimHeader{
+		Type:       *ClaimTypeAuthorizeKSignBabyJub,
+		Dest:       ClaimRecipSelf,
+		Expiration: false,
+		Version:    false}
+)
+
+// EntrierMetadata is an intefrace that extends Entrier with a function that
+// returns the claim metadata.
+type EntrierMetadata interface {
+	merkletree.Entrier
+	Metadata() *Metadata
+}
+
+// Metadata is a header and generic (some optional) values of a claim.
+type Metadata struct {
+	header     ClaimHeader
+	Dest       *core.ID
+	Expiration int64
+	Version    uint32
+	RevNonce   uint32
+}
+
+// NewMetadata creates a new Metadata with a specific header.
+func NewMetadata(header ClaimHeader) Metadata {
+	return Metadata{header: header}
+}
+
+// Header returns the header from the metadata.
+func (m *Metadata) Header() ClaimHeader {
+	return m.header
+}
+
+// Type returns the claim type from the header in the metadata.
+func (m *Metadata) Type() ClaimType {
+	return m.header.Type
+}
+
+// Marshal the Metadata into an entry
+func (m *Metadata) Marshal(e *merkletree.Entry) {
+	m.header.Marshal(e)
+	index := e.Index()
+	value := e.Value()
+	switch m.header.Dest {
+	case ClaimRecipSelf:
+	case ClaimRecipIdenIndex:
+		copy(index[1][:], m.Dest[:])
+	case ClaimRecipIdenValue:
+		copy(value[1][:], m.Dest[:])
+	default:
+		panic(fmt.Sprintf("Unexpected header.Dest %v", m.header.Dest))
+	}
+	if m.header.Version {
+		binary.BigEndian.PutUint32(index[0][ClaimTypeLen+ClaimFlagsLen:], m.Version)
+	}
+	if m.header.Expiration {
+		binary.BigEndian.PutUint64(value[0][32/8:], uint64(m.Expiration))
+	}
+	binary.BigEndian.PutUint32(value[0][:], m.RevNonce)
+}
+
+// Unmarshal the Metadata from an entry
+func (m *Metadata) Unmarshal(e *merkletree.Entry) {
+	m.header.Unmarshal(e)
+	index := e.Index()
+	value := e.Value()
+	if m.header.Dest != ClaimRecipSelf {
+		m.Dest = &core.ID{}
+	}
+	switch m.header.Dest {
+	case ClaimRecipSelf:
+	case ClaimRecipIdenIndex:
+		copy(m.Dest[:], index[1][:])
+	case ClaimRecipIdenValue:
+		copy(m.Dest[:], value[1][:])
+	default:
+		panic(fmt.Sprintf("Unexpected header.Dest %v", m.header.Dest))
+	}
+	if m.header.Version {
+		m.Version = binary.BigEndian.Uint32(index[0][ClaimTypeLen+ClaimFlagsLen:])
+	}
+	if m.header.Expiration {
+		m.Expiration = int64(binary.BigEndian.Uint64(value[0][32/8:]))
```

Comment:
Same as previous comment

Reply:
I have modified the code to use the `ClaimRevNonceLen` constant